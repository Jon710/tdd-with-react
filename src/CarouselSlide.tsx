import { ComponentPropsWithRef, ReactNode } from "react";
import styled from "styled-components";

const DEFAULT_IMG_HEIGHT = "500px";

export type CarouselSlideProps = {
  imgUrl?: string;
  /** @default "500px" */
  imgHeight?: string | number;
  description?: ReactNode;
  attribution?: ReactNode;
} & ComponentPropsWithRef<"figure">;

type ImgComponentProps = {
  $height?: CarouselSlideProps["imgHeight"];
};

const ScaledImg = styled.img<ImgComponentProps>`
  object-fit: cover;
  width: 100%;
  height: ${({ $height }) =>
    typeof $height === "number" ? `${$height}px` : $height};
`;

function CarouselSlide({
  imgUrl,
  imgHeight = DEFAULT_IMG_HEIGHT,
  description,
  attribution,
  ...rest
}: CarouselSlideProps) {
  return (
    <figure {...rest}>
      <ScaledImg src={imgUrl} $height={imgHeight} />
      <figcaption data-testid="caption">
        <strong>{description}</strong> {attribution}
      </figcaption>
    </figure>
  );
}

export default CarouselSlide;
