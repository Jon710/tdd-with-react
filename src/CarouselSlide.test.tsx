import { render, screen } from "@testing-library/react";
import CarouselSlide from "./CarouselSlide";

describe("CarouselSlide component", () => {
  it("matches the snapshot", () => {
    render(<CarouselSlide />);
    expect(screen.getByRole("figure")).toMatchSnapshot();
  });

  it("renders a figure element", () => {
    render(<CarouselSlide />);
    const figure = screen.getByRole("figure");
    const img = screen.getByRole("img");
    const figCaption = screen.getByTestId("caption");

    expect(figure).toContainElement(img);
    expect(figure).toContainElement(figCaption);
  });

  it("passes imgUrl thru to the <img>", () => {
    const imgUrl = "https://example.com/image.jpg";
    render(<CarouselSlide imgUrl={imgUrl} />);
    expect(screen.getByRole("img")).toHaveAttribute("src", imgUrl);
  });

  it("uses `description` and `attribution` as the caption", () => {
    const props = {
      description: "A beautiful image",
      attribution: "John Doe",
    };

    render(<CarouselSlide {...props} />);

    const figCaption = screen.getByTestId("caption");
    expect(figCaption).toHaveTextContent(
      `${props.description} ${props.attribution}`
    );
  });

  it("passes other props thru to the figure", () => {
    const props = { className: "my-slide", "data-test-name": "my slide" };

    render(<CarouselSlide {...props} />);

    const figure = screen.getByRole("figure");
    expect(figure).toHaveClass(props.className);
    expect(figure).toHaveAttribute("data-test-name", props["data-test-name"]);
  });
});
