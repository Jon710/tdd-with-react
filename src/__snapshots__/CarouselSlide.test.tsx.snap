// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CarouselSlide component > matches the snapshot 1`] = `
<figure>
  <img
    class="sc-Qotzb gnprED"
  />
  <figcaption
    data-testid="caption"
  >
    <strong />
     
  </figcaption>
</figure>
`;
